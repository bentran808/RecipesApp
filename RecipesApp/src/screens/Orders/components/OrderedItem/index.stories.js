import { storiesOf } from '@storybook/react-native';
import React from 'react';
import CenterView from 'storybook/stories/CenterView';
import OrderedItem from '.';

storiesOf('Ordered Item', module)
  .addDecorator((getStory) => <CenterView>{getStory()}</CenterView>)
  .add('default', () => (
    <OrderedItem
      item={{
        items: [
          {
            name: 'Oatmeal Cookies',
            quantity: 1
          },
          {
            name: 'Triple Berry Smoothie',
            quantity: 2
          }
        ],
        total: 65,
        createdAt: '06 May 2022, 4:30PM'
      }}
    />
  ));
