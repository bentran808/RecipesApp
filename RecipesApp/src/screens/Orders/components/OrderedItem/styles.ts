import { StyleSheet } from 'react-native';
import { width } from 'theme';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 20,
    paddingVertical: 10,
    width: width
  },
  info: {
    maxWidth: 200
  },
  time: {
    fontSize: 12,
    marginTop: 5
  },
  customBtn: {
    width: 130,
    borderRadius: 5,
    marginTop: 10,
    marginLeft: 20
  }
});

export default styles;
