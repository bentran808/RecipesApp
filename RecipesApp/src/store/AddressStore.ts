import { toJS } from 'mobx';
import { types, destroy, SnapshotOut, getRoot, applySnapshot } from 'mobx-state-tree';
import { RootStore } from 'store/store';

const AddressEntry = types
  .model('AddressEntry', {
    id: types.optional(types.identifierNumber, 0),
    type: types.optional(types.string, ''),
    address: types.optional(types.string, '')
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).address.removeItem(self);
    },
    edit(type: string, address: string) {
      self.type = type;
      self.address = address;
    }
  }));

export type AddressModel = SnapshotOut<typeof AddressEntry>;

const AddressStore = types
  .model('', {
    items: types.array(AddressEntry),
    itemUsing: types.optional(AddressEntry, {})
  })
  .volatile(() => ({
    address: ''
  }))
  .actions((self) => ({
    setAddress(address: string) {
      self.address = address;
    },
    addNewItem(type: string, address: string) {
      const id = self.items.reduce((maxId, item) => Math.max(item.id, maxId), -1) + 1;
      self.items.push({ id, type, address });
    },
    removeItem(item: AddressModel) {
      destroy(item);
    },
    existAddress(address?: AddressModel) {
      return self.items.find((item) => item.id === address?.id);
    },
    setItemUsing(item: AddressModel) {
      applySnapshot(self, {
        ...self,
        itemUsing: item
      });
    }
  }))
  .views((self) => ({
    get addresses() {
      return toJS(self.items);
    }
  }));

export default AddressStore;
