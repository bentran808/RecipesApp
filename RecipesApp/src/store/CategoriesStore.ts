import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';
import { transformCategories } from 'utils';
import { RecipesEntry } from './RecipesStore';

export const CategoriesEntry = types.model('CategoriesEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string,
  recipes: types.maybe(types.array(RecipesEntry))
});

export type CategoryModel = SnapshotOut<typeof CategoriesEntry>;

const CategoriesStore = types
  .model('CategoriesStore', {
    lists: types.optional(types.array(CategoriesEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get categories() {
      return self.lists;
    },
    get categoriesJS() {
      return transformCategories(toJS(this.categories));
    }
  }))
  .actions((self) => ({
    fetchCategories: flow(function* fetchCategories() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchCategoriesRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        console.log(error);
        self.state = 'error';
      }
    })
  }));

export default CategoriesStore;
