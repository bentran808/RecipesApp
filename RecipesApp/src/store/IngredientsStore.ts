import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';

export const IngredientsEntry = types.model('IngredientsEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string
});

export type IngredientModel = SnapshotOut<typeof IngredientsEntry>;

const IngredientsStore = types
  .model('IngredientsStore', {
    lists: types.optional(types.array(IngredientsEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get ingredients() {
      return self.lists;
    },
    get ingredientsJS() {
      return toJS(this.ingredients);
    }
  }))
  .actions((self) => ({
    fetchIngredients: flow(function* fetchIngredients() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchIngredientsRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    })
  }));

export default IngredientsStore;
