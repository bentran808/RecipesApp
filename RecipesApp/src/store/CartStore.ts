import { toJS } from 'mobx';
import { applySnapshot, destroy, getRoot, SnapshotOut, types } from 'mobx-state-tree';
import { RecipeModel, RecipesEntry } from 'store/RecipesStore';
import { RootStore } from 'store/store';

export const CartEntry = types
  .model('CartEntry', {
    item: RecipesEntry,
    quantity: types.optional(types.number, 1),
    inCart: types.optional(types.boolean, true)
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).cart.removeItem(self);
    },
    increase(quantity?: number) {
      self.quantity = quantity ? self.quantity + quantity : ++self.quantity;
    },
    decrease() {
      self.quantity === 1 ? this.remove() : --self.quantity;
    },
    toggle() {
      self.inCart = !self.inCart;
    }
  }));

const BillEntry = types.model('BillEntry', {
  type: types.optional(types.string, ''),
  name: types.optional(types.string, ''),
  code: types.optional(types.string, ''),
  price: types.optional(types.number, 0)
});

const OrderedEntry = types.model('OrderedEntry', {
  items: types.array(
    types.model({
      name: types.optional(types.string, ''),
      quantity: types.optional(types.number, 0)
    })
  ),
  total: types.number,
  createdAt: types.string
});

export type CartModel = SnapshotOut<typeof CartEntry>;
export type BillModel = SnapshotOut<typeof BillEntry>;
export type OrderedModel = SnapshotOut<typeof OrderedEntry>;

const CartStore = types
  .model('CartStore', {
    items: types.array(CartEntry),
    billItems: types.array(BillEntry),
    orderedItems: types.array(OrderedEntry)
  })
  .views((self) => ({
    get recipes() {
      return toJS(self.items);
    },
    get recipesInCart() {
      return this.recipes.filter((item) => item.inCart);
    },
    get inCartCount() {
      return this.recipesInCart.length;
    },
    get total() {
      return this.recipesInCart.reduce((sum, item) => sum + item.quantity * 20, 0);
    },
    get billDetails() {
      return toJS(self.billItems);
    },
    get toPay() {
      return this.billDetails.reduce(
        (sum, item) => (item.type === 'discount' ? sum - item.price : sum + item.price),
        0
      );
    },
    get orderedList() {
      return toJS(self.orderedItems);
    }
  }))
  .volatile(() => ({
    discountInput: '',
    invalidCoupon: false
  }))
  .actions((self) => ({
    addToCart(item: RecipeModel, quantity?: number) {
      const itemInCart = this.hasInCart(item);
      itemInCart ? itemInCart.increase(quantity) : self.items.push({ item, quantity });
    },
    removeItem(item: CartModel) {
      destroy(item);
    },
    hasInCart(recipe: RecipeModel) {
      return self.items.find((item) => item.item.id === recipe.id);
    },
    setDiscountInput(text: string) {
      if (!text) {
        self.invalidCoupon = false;
      }
      self.discountInput = text;
    },
    applyCoupon(item: BillModel) {
      const existItem = self.billItems.findIndex((bill) => bill.type === 'discount');
      if (existItem > -1) {
        self.billItems.splice(existItem, 1, item);
      } else {
        self.billItems.push(item);
      }
    },
    payment(order: OrderedModel) {
      self.orderedItems.unshift(order);
      applySnapshot(self, {
        ...self,
        items: [],
        billItems: []
      });
    },
    isInvalidCoupon(status: boolean) {
      self.invalidCoupon = status;
    }
  }));

export default CartStore;
