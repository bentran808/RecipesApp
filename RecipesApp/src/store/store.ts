import { Instance, types } from 'mobx-state-tree';
import RecipesStore from 'store/RecipesStore';
import CategoriesStore from './CategoriesStore';

const RootStore = types.model('RootStore', {
  categories: CategoriesStore,
  recipes: RecipesStore
});

const store = RootStore.create({
    categories: {},
    recipes: {}
});

export type RootStoreModel = Instance<typeof RootStore>;

export default store;
