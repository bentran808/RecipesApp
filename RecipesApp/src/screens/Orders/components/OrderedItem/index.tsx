import AppStyles from 'AppStyles';
import Button from 'components/Button';
import React from 'react';
import isEqual from 'react-fast-compare';
import { Text, View } from 'react-native';
import { OrderedModel } from 'store/CartStore';
import styles from './styles';

type Props = {
  item: OrderedModel;
};

const OrderedItem = ({ item }: Props) => {
  const title = item.items
    .map((orderedItem) => `${orderedItem.name} x ${orderedItem.quantity}`)
    .join(', ');

  return (
    <View
      style={{
        paddingBottom: 15
      }}
    >
      <View style={styles.container}>
        <View style={styles.info}>
          <Text style={AppStyles.textBold}>{title}</Text>
          <Text style={[AppStyles.subText, styles.time]}>{item.createdAt}</Text>
        </View>
        <Text>${item.total.toFixed(2)}</Text>
      </View>
      <Button
        testID="reorderBtn"
        type="outlined"
        bold
        title="REORDER"
        paddingHorizontal={30}
        paddingVertical={5}
        color="red"
        borderColor="red"
        onPress={() => {}}
        style={styles.customBtn}
      />
    </View>
  );
};

export default React.memo(OrderedItem, isEqual);
