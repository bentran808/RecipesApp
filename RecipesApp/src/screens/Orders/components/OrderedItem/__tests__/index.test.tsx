import { StoreProvider } from 'context';
import React from 'react';
import renderer from 'react-test-renderer';
import store from 'store/store';
import OrderedItem from '..';

describe('Ordered Item Component', () => {
  const props = {
    item: {
      items: [
        {
          name: 'test',
          quantity: 2
        }
      ],
      total: 40,
      createdAt: ''
    }
  };

  test('should render correctly', () => {
    const tree = renderer
      .create(
        <StoreProvider value={store}>
          <OrderedItem {...props} />
        </StoreProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
