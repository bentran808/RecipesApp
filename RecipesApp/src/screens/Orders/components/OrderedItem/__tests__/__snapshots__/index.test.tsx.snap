// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ordered Item Component should render correctly 1`] = `
<View
  style={
    Object {
      "paddingBottom": 15,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 20,
        "paddingVertical": 10,
        "width": 750,
      }
    }
  >
    <View
      style={
        Object {
          "maxWidth": 200,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        test x 2
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "gray",
            },
            Object {
              "fontSize": 12,
              "marginTop": 5,
            },
          ]
        }
      >
        
      </Text>
    </View>
    <Text>
      $
      40.00
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="reorderBtn"
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 100,
            "marginHorizontal": 10,
          },
          Object {
            "borderColor": "red",
            "borderWidth": 1,
            "paddingHorizontal": 30,
            "paddingVertical": 5,
          },
          Object {
            "borderRadius": 5,
            "marginLeft": 20,
            "marginTop": 10,
            "width": 130,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "red",
              "fontSize": 14,
              "fontWeight": "bold",
              "textAlign": "center",
            },
          ]
        }
      >
        REORDER
      </Text>
    </View>
  </View>
</View>
`;
